File: site/03-Stacks-and-Queues/js/balanced_parentheses.js
function isBalanced(s) {
    const stack = [];
    for (let i = 0; i < s.length; i++) {
        const c = s[i];
        if (c === '(' || c === '{' || c === '[') {
            stack.push(c);
        } else {
            if (stack.length === 0) return false;
            const top = stack[stack.length - 1];
            if ((c === ')' && top === '(') ||
                (c === '}' && top === '{') ||
                (c === ']' && top === '[')) {
                stack.pop();
            } else {
                return false;
            }
        }
    }
    return stack.length === 0;
}

// Test cases
const test1 = "{[()]}";
const test2 = "{[(])}";
console.log(`Testing: "${test1}" -> ${isBalanced(test1) ? "Balanced" : "Not Balanced"}`);
console.log(`Testing: "${test2}" -> ${isBalanced(test2) ? "Balanced" : "Not Balanced"}`);

File: site/03-Stacks-and-Queues/java/BalancedParentheses.java
import java.util.Stack;

public class BalancedParentheses {
    public static boolean isBalanced(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) return false;
                char top = stack.peek();
                if ((c == ')' && top == '(') ||
                    (c == '}' && top == '{') ||
                    (c == ']' && top == '[')) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    public static void main(String[] args) {
        String test1 = "{[()]}";
        String test2 = "{[(])}";
        
        System.out.println("Testing: \"" + test1 + "\" -> " + (isBalanced(test1) ? "Balanced" : "Not Balanced"));
        System.out.println("Testing: \"" + test2 + "\" -> " + (isBalanced(test2) ? "Balanced" : "Not Balanced"));
    }
}
